### PR TITLE
stringFormat extractionFn should be able to return null on null values (Fix for #2706)

### PR DESCRIPTION
**Pull request by navis**
Fix for #2706

### CODE REVIEW

**Comment by navis** [comment]
test fail is not related to this

**Comment by gianm** [comment]
@navis you can close and re-open to get tests to re-run if the failure is not related.

**Comment by gianm** [comment]
@navis looks good, could you please include docs too though?

File: processing/src/main/java/io/druid/query/extraction/StringFormatExtractionFn.java [hunk 1]
```diff
@@ -33,14 +33,26 @@
...
+      @JsonProperty("format") String format,
+      @JsonProperty("nullAsNull") boolean nullAsNull,
+      @JsonProperty("nullAsEmpty") boolean nullAsEmpty
   )
   {
     Preconditions.checkArgument(!Strings.isNullOrEmpty(format), "format string should not be empty");
+    Preconditions.checkArgument(!(nullAsNull && nullAsEmpty), "nullAsNull and nullAsEmpty cannot be both true");
     this.format = format;
+    this.nullAsNull = nullAsNull;
+    this.nullAsEmpty = nullAsEmpty;
```
Comment (himanshug) [inline processing/src/main/java/io/druid/query/extraction/StringFormatExtractionFn.java]: why do we need 2 boolean variables? can't we have just one variable which if set to true then maps "null to null" and "null to empty" when set to false.

Comment (gianm) [inline processing/src/main/java/io/druid/query/extraction/StringFormatExtractionFn.java]: there are 3 behaviors (see the original issue #2706),

- treat `null` as `"null"`, `stringFormat("[%s]")` returns `"[null]"` (current)
- treat `null` as `null`, `stringFormat("[%s]")` returns `null` (nullAsNull)
- treat `null` as `""`, `stringFormat("[%s]")` returns `"[]"` (nullAsEmpty)

Comment (jon-wei) [inline processing/src/main/java/io/druid/query/extraction/StringFormatExtractionFn.java]: Would that be better handled with a 3-value enum for "null handling behavior", vs. using booleans?

Comment (gianm) [inline processing/src/main/java/io/druid/query/extraction/StringFormatExtractionFn.java]: an enum sounds good to me.

**Comment by fjy** [comment]
@navis any chance of finishing this up?

**Comment by navis** [comment]
@fjy changed to enum

**Comment by gianm** [comment]
unrelated test failure, bouncing.

```
Running io.druid.indexing.kafka.KafkaIndexTaskTest
Tests run: 25, Failures: 0, Errors: 6, Skipped: 0, Time elapsed: 219.436 sec <<< FAILURE! - in io.druid.indexing.kafka.KafkaIndexTaskTest
testRunOneTaskTwoPartitions[buildV9Directly = true](io.druid.indexing.kafka.KafkaIndexTaskTest)  Time elapsed: 7.834 sec  <<< ERROR!
org.I0Itec.zkclient.exception.ZkTimeoutException: Unable to connect to zookeeper server within timeout: 6000
	at org.I0Itec.zkclient.ZkClient.connect(ZkClient.java:1223)
	at org.I0Itec.zkclient.ZkClient.<init>(ZkClient.java:155)
	at org.I0Itec.zkclient.ZkClient.<init>(ZkClient.java:129)
	at kafka.utils.ZkUtils$.createZkClientAndConnection(ZkUtils.scala:89)
	at kafka.utils.ZkUtils$.apply(ZkUtils.scala:71)
	at kafka.server.KafkaServer.initZk(KafkaServer.scala:278)
	at kafka.server.KafkaServer.startup(KafkaServer.scala:168)
	at io.druid.indexing.kafka.test.TestBroker.start(TestBroker.java:69)
	at io.druid.indexing.kafka.KafkaIndexTaskTest.setUp(KafkaIndexTaskTest.java:245)
```

**Comment by gianm** [comment]
:+1: after travis passes & commits squashed

**Comment by navis** [comment]
squashed

File: processing/src/main/java/io/druid/query/extraction/StringFormatExtractionFn.java [hunk 2]
```diff
@@ -32,15 +33,47 @@
  */
 public class StringFormatExtractionFn extends DimExtractionFn
 {
+  public static enum NullHandling
+  {
+    NULLSTRING,
```
Comment (jon-wei) [inline processing/src/main/java/io/druid/query/extraction/StringFormatExtractionFn.java]: nit: can we put underscores in the names of the enum values?

Comment (gianm) [inline processing/src/main/java/io/druid/query/extraction/StringFormatExtractionFn.java]: I assume the idea here is to make it easy for `NullHandling.valueOf(value.toUpperCase())` to work on "nullString"

**Comment by vogievetsky** [comment]
looks great thank you @navis

File: processing/src/main/java/io/druid/query/extraction/StringFormatExtractionFn.java (hunk 2 above)
Comment (jon-wei) [inline processing/src/main/java/io/druid/query/extraction/StringFormatExtractionFn.java]: I see, nm, that's fine then

File: processing/src/main/java/io/druid/query/extraction/StringFormatExtractionFn.java [hunk 3]
```diff
@@ -32,15 +33,47 @@
...
+      @JsonProperty("format") String format,
+      @JsonProperty("nullHandling") NullHandling nullHandling
   )
   {
     Preconditions.checkArgument(!Strings.isNullOrEmpty(format), "format string should not be empty");
     this.format = format;
+    this.nullHandling = nullHandling == null ? NullHandling.NULLSTRING : nullHandling;
+  }
+
+  public StringFormatExtractionFn(String format, String nullHandling)
```
Comment (jon-wei) [inline processing/src/main/java/io/druid/query/extraction/StringFormatExtractionFn.java]: Are these extra constructors necessary? Seems like they're only used within the included test file, the forValue() and toValue() functions are enough for serde to work

Comment (navis) [inline processing/src/main/java/io/druid/query/extraction/StringFormatExtractionFn.java]: done

**Comment by gianm** [comment]
#2253 

```
Tests run: 5, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 67.136 sec <<< FAILURE! - in io.druid.curator.announcement.AnnouncerTest
testSanity(io.druid.curator.announcement.AnnouncerTest)  Time elapsed: 61.074 sec  <<< ERROR!
java.lang.Exception: test timed out after 60000 milliseconds
	at sun.misc.Unsafe.park(Native Method)
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedNanos(AbstractQueuedSynchronizer.java:1037)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.tryAcquireSharedNanos(AbstractQueuedSynchronizer.java:1328)
	at java.util.concurrent.CountDownLatch.await(CountDownLatch.java:277)
	at org.apache.curator.test.Timing.awaitLatch(Timing.java:120)
	at io.druid.curator.announcement.AnnouncerTest.testSanity(AnnouncerTest.java:101)
```

**Comment by gianm** [comment]
:+1:

**Comment by drcrallen** [comment]
The apply needs more enforcement, from javadocs:


```java
  /**
   * The "extraction" function.  This should map a value into some other String value.
   * <p>
   * In order to maintain the "null and empty string are equivalent" semantics that Druid provides, the
   * empty string is considered invalid output for this method and should instead return null.  This is
   * a contract on the method rather than enforced at a lower level in order to eliminate a global check
   * for extraction functions that do not already need one.
   *
   * @param value the original value of the dimension
   *
   * @return a value that should be used instead of the original
   */
  public String apply(Object value);
```

The return value of apply needs `Strings.emptyToNull`

**Comment by gianm** [comment]
good point @drcrallen

**Comment by navis** [comment]
@drcrallen nice catch. fixed and added test case.

**Comment by fjy** [comment]
@navis can you squash commits and ill merge

**Comment by navis** [comment]
squahsed

**Comment by fjy** [comment]
:+1:

**Pull request merged**